controller/lib/hal/hal_stm32_regs.h

```diff
@@ -0,0 +1,444 @@
+    REG teif6 : 1; // transfer error (TE) flag
+    REG gif7 : 1;  // global interrupt flag
+    REG tcif7 : 1; // transfer complete (TC) flag
+    REG htif7 : 1; // half transfer (HT) flag
+    REG teif7 : 1; // transfer error (TE) flag
+    REG rsvd : 4;
+  } intStat; // interrupt status register
+
+  struct {
+    REG gif1 : 1;  // global interrupt flag
+    REG tcif1 : 1; // transfer complete (TC) flag
+    REG htif1 : 1; // half transfer (HT) flag
+    REG teif1 : 1; // transfer error (TE) flag
+    REG gif2 : 1;  // global interrupt flag
+    REG tcif2 : 1; // transfer complete (TC) flag
+    REG htif2 : 1; // half transfer (HT) flag
+    REG teif2 : 1; // transfer error (TE) flag
+    REG gif3 : 1;  // global interrupt flag
+    REG tcif3 : 1; // transfer complete (TC) flag
+    REG htif3 : 1; // half transfer (HT) flag
+    REG teif3 : 1; // transfer error (TE) flag
+    REG gif4 : 1;  // global interrupt flag
+    REG tcif4 : 1; // transfer complete (TC) flag
+    REG htif4 : 1; // half transfer (HT) flag
+    REG teif4 : 1; // transfer error (TE) flag
+    REG gif5 : 1;  // global interrupt flag
+    REG tcif5 : 1; // transfer complete (TC) flag
+    REG htif5 : 1; // half transfer (HT) flag
+    REG teif5 : 1; // transfer error (TE) flag
+    REG gif6 : 1;  // global interrupt flag
+    REG tcif6 : 1; // transfer complete (TC) flag
+    REG htif6 : 1; // half transfer (HT) flag
+    REG teif6 : 1; // transfer error (TE) flag
+    REG gif7 : 1;  // global interrupt flag
+    REG tcif7 : 1; // transfer complete (TC) flag
+    REG htif7 : 1; // half transfer (HT) flag
+    REG teif7 : 1; // transfer error (TE) flag
+    REG rsvd : 4;
+  } intClr; // interrupt flag clear register
+  struct {
+    struct {
+      REG enable : 1;   // channel enable
+      REG tcie : 1;     // transfer complete interrupt enable
+      REG htie : 1;     // half transfer interrupt enable
+      REG teie : 1;     // transfer error interrupt enable
+      REG dir : 1;      // data xfer direction 0: per->mem, 1: mem->per
+      REG circular : 1; // circular mode
+      REG perInc : 1;   // peripheral increment mode
+      REG memInc : 1;   // memory increment mode
+      REG psize : 2;    // peripheral size 0b00 - 8bits, 0b10 - 32bits
+      REG msize : 2;    // memory size 0b00 - 8bits, 0b10 - 32bits
+      REG priority : 2; // priority level 0b00 - low, 0b11 - high
+      REG mem2mem : 1;  // memory-to-memory mode
+      REG rsvd : 17;
+    } config;  // channel x configuration register
+    REG count; // channel x number of data to transfer register
+    REG pAddr; // channel x peripheral address register
+    REG mAddr; // channel x memory address register
+    REG rsvd;  // reserved
+  } channel[7];
+  REG rsvd[5]; // reserved
+  struct {
+    REG c1s : 4;
+    REG c2s : 4;
+    REG c3s : 4;
+    REG c4s : 4;
+    REG c5s : 4;
+    REG c6s : 4;
+    REG c7s : 4;
+    REG rsvd : 4;
+  } chanSel; // channel selection register
```

Comment:
Like with the interrupt registers above, I think it's better if we can have the channel be a variable and not hard code things, so I don't think the bit-flags work well for this register.  A function to set the value based on channel number would be nicer.

Reply:
I have pondered about this also. I have solve it by decoupling channel selection and bit clearing from UART DMA configuration.

Channel selection happens in DMA.init() method. Since UART3 and UART2 are fixed and available on a single channel configuration, there is no sense implementing the whole matrix in the datasheet. Since they are hardcoded in hardware, I think it's OK to hardcode the same configuration in software. See `uart_dma.h`  `DMACtrl::init()`

Regarding flag clearing register I have moved the related code to the interrupt handlers for matching registers. Since interrupt is specific, say `DMA1_CH2_ISR` it's ok to clear CH2 related flags in it. See `uart_dma.cpp:241`.

`UART_DMA` is abstract enough to support several UARTS. What do you think?